DisplayListRecorder fails to record state change before multiple commands

kkinnunen-apple opened the pull request:
#### b9e5c9199441501375c71988e4efca5f487905f2
<pre>
DisplayListRecorder fails to record state change before multiple commands
<a href="https://bugs.webkit.org/show_bug.cgi?id=253693">https://bugs.webkit.org/show_bug.cgi?id=253693</a>
rdar://problem/106542943

Reviewed by Simon Fraser.

Problems:
1. Multiple commands that use or might use the state are missing the
   appendStateChangeItemIfNecessary() that would write the state item.
2. Recorder::clipToImageBuffer() does not maintain the state needed to
   return the correct GraphicsContext::clipBounds().

Fix 1. by applying appendStateChangeItemIfNecessary() conservatively.
This is likely not a perf problem, since all the clip commands
would anyway need some future draw command, and those would append
the state change at that point.

Fix 2. by updating the clip bounds, similar to other cases which
update the clip bounds.

The video drawing commands are not testable currently, as MediaPlayer
cannot be instantiated as a mock and VideoFrame command is not implemented.

Auxiliary problems:
The unfixed, buggy behavior of Recorder::clipToPath() would invert buggy
behavior of BorderPainter.

This would lead to passing
imported/w3c/web-platform-tests/css/css-ui/compute-kind-widget-generated/kind-of-widget-fallback-input-button-border-block-end-color-001.html
and similar tests.

The example of the problem, test above would compare two renderings:

Reference (kind-of-widget-fallback-input-button-border-block-end-color-001-expected.html)
    &lt;input id=&quot;button-input&quot; type=&quot;button&quot; value=&quot;input-button&quot;&gt;

vs:

Test (kind-of-widget-fallback-input-button-border-block-end-color-001.html)
    &lt;input id=&quot;button-input&quot; type=&quot;button&quot; value=&quot;input-button&quot;&gt;

    const prop = &quot;border-block-end-color&quot;;
    for (const el of elements) {
        el.style.setProperty(prop, getComputedStyle(el).getPropertyValue(prop));
    }

The BorderPainter would compare the colors of each border, and run the
fast path if all the colors match.

For the reference, the fast path was taken.

For the test, because of the BorderPainter bug, the colors would not match. The
BorderPainter would take slow path that would set up complex clip path
with antialiasing.

Because the Recorder::clipToPath would incorrectly skip setting the antialiasing
state, two corner pixels of the rounded rect would be included in the rendering.
This would match the unclipped fast path rendering of the reference.

Once the clipToPath() bug was fixed, it would show two-pixel difference due
to the clipToPath being correctly cliping antialiased path vs the reference
using the fast path, unclipped.

Fix by fixing BorderPainter to compare the border colors always without
the &quot;semantic&quot; bit.

This makes the test take the same fast path as the reference, and thus the
rendering is identical.

Also in a best-effort manner, fix one other color compare case to ignore
the semantic bit. It is not consistent that one code-path would ignore
the semantic bit while others would be continue using it.

* Source/WebCore/platform/graphics/displaylists/DisplayListRecorder.cpp:
(WebCore::DisplayList::Recorder::clip):
(WebCore::DisplayList::Recorder::clipOut):
(WebCore::DisplayList::Recorder::clipPath):
(WebCore::DisplayList::Recorder::clipToImageBuffer):
(WebCore::DisplayList::Recorder::paintFrameForMedia):
(WebCore::DisplayList::Recorder::paintVideoFrame):
* Tools/TestWebKitAPI/Tests/WebCore/DisplayListRecorderTests.cpp:
(TestWebKitAPI::createTestPath):
(TestWebKitAPI::createTestImageBuffer):
(TestWebKitAPI::checkEqualState):
* Source/WebCore/rendering/BorderPainter.cpp:
 (WebCore::BorderPainter::paintSides):
 (WebCore::BorderPainter::paintTranslucentBorderSides):

Canonical link: <a href="https://commits.webkit.org/261618@main">https://commits.webkit.org/261618@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/dd042adf409f52963ada3cbc217df38fd993996a

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/112171 "44 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/21305 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/90/builds/797 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/87/builds/3971 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/120826 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/116234 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/22653 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/12509 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/86/builds/4586 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/117934 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/16836 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/100010 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/105232 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/98793 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/88/builds/552 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/45839 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/13715 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/89/builds/587 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/93709 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/14393 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/84/builds/9988 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/19761 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/52586 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/8100 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/16186 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

shallawa commented on Source/WebCore/platform/graphics/displaylists/DisplayListRecorder.cpp:
```diff
@@ -531,22 +531,26 @@ void Recorder::drawControlPart(ControlPart& part, const FloatRoundedRect& border
 
 void Recorder::clip(const FloatRect& rect)
 {
+    appendStateChangeItemIfNecessary(); // Conservative: we do not know if the clip application might use state such as antialiasing.
```
I think we have 63 display list items. Currently, there are 36 display list items call to this function. New calls to this function have been added recently. By adding 7 calls by this PR, we will have 43 calls.

I think we should be more conservative in calling this function. How about adding a new function like this:

```
void Recorder::appendStateChangeItemIfNecessary(GraphicsContextState::Change change)
```
or
```
void Recorder::appendStateChangeItemIfNecessary(GraphicsContextState::ChangeFlags changes)
```
This function will call the existing one if the current state changes the context with `change` or any of `changes`. The above function can call it like this
```
appendStateChangeItemIfNecessary(GraphicsContextState::Change::ShouldAntialias);
```

kkinnunen-apple replied:
So `appendStateChangeItemIfNecessary()` of course already consults the change flags and only sends state changes that have actually occurred.

> I think we have 63 display list items. Currently, there are 36 display list items call to this function. New calls to this function have been added recently. By adding 7 calls by this PR, we will have 43 calls.

In general, almost all calls should sync the state. I don't think there's many calls that draw that do not need the exactly correct draw state. To me it just sounds natural and I'm not terribly worried the increase of calls. I'm more worried about the calls that don't have the state sync -- why are they correct?? 

If I understand correctly, you're concerned that if there are pending change tags, say, `GraphicsContextState::Change::ShouldAntialias` and `GraphicsContextState::Change::ShouldSmoothFonts`, and `clip()` only needs `ShouldAntialias`, then we shouldn't send `ShouldSmoothFonts `?

I'm not sure that is an optimisation at all. Generally the `appendStateChangeItemIfNecessary` is an optimisation for batching consecutive changes to the state. In other words: the painting, typically, does not set state that is not needed. Thus for CSS painting, vast majority of the state that is set is also used. 

Sending all the set state in itself is really lightweight operation.

Canvas could change state that is not used, but that, too, never have shown up in our benchmarks -- this function is does not seem to do redundant work. We also don't have that optimisation for non-gpup: current in-process context does set all requested state eagerly, even  though the benefit should be the same for gpu and non-gpu.

If the state is set, it is likely to be used. There's no evidence that we set the state redundantly. This would mean that more fine-grained state setting would be just splitting the state setting into two state setting functions, something that probably in itself is a regression (two IPC commands vs one). The only gain would be if some state is set redundantly, and we'd actually skip using it altogether.

It is also very easy to get wrong -- very much attention would need to be put to re-learn all commands and which state they depend on.

Maybe making that change would need some data to back it as an optimisation?

shallawa replied:
> If I understand correctly, you're concerned that if there are pending change tags, say, GraphicsContextState::Change::ShouldAntialias and GraphicsContextState::Change::ShouldSmoothFonts, and clip() only needs ShouldAntialias, then we shouldn't send ShouldSmoothFonts ?

No I did not mean that. Consider these two cases:

// Case 1: context state does not have ShouldAntialias change
```
context.setFillColor(...)
context.clip(...)
```

// Case 2: context state does have ShouldAntialias change
```
context.setShouldAntialias(...)
context.clip(...)
```

For case 1, we may not need to `appendStateChangeItemIfNecessary()` because the pending context state does not have anything that will affect clipping. However for case 2, we need to commit the pending context state before applying the clipping.

So if we make `Recorder::clip()` calls 

```
appendStateChangeItemIfNecessary(GraphicsContextState::Change::ShouldAntialias);
```
It will not call `appendStateChangeItemIfNecessary()` for case 1 but will call it for case 2.

shallawa replied:
Pleas note that my suggestion is to make DisplayList::Recorder have two versions of `appendStateChangeItemIfNecessary()`; something like that:
```
void appendStateChangeItemIfNecessary();
void appendStateChangeItemIfNecessary(GraphicsContextState::ChangeFlags changes);
```

The implementation of the second version can be something like this:

```
void Recorder::appendStateChangeItemIfNecessary(GraphicsContextState::ChangeFlags changes)
{
    auto& state = currentState().state;
    if (!state.changes().containsAny(changes)))
        return;
    appendStateChangeItemIfNecessary();
}
```

kkinnunen-apple committed:
DisplayListRecorder fails to record state change before multiple commands
https://bugs.webkit.org/show_bug.cgi?id=253693
rdar://problem/106542943

Reviewed by Simon Fraser.

Problems:
1. Multiple commands that use or might use the state are missing the
   appendStateChangeItemIfNecessary() that would write the state item.
2. Recorder::clipToImageBuffer() does not maintain the state needed to
   return the correct GraphicsContext::clipBounds().

Fix 1. by applying appendStateChangeItemIfNecessary() conservatively.
This is likely not a perf problem, since all the clip commands
would anyway need some future draw command, and those would append
the state change at that point.

Fix 2. by updating the clip bounds, similar to other cases which
update the clip bounds.

The video drawing commands are not testable currently, as MediaPlayer
cannot be instantiated as a mock and VideoFrame command is not implemented.

Auxiliary problems:
The unfixed, buggy behavior of Recorder::clipToPath() would invert buggy
behavior of BorderPainter.

This would lead to passing
imported/w3c/web-platform-tests/css/css-ui/compute-kind-widget-generated/kind-of-widget-fallback-input-button-border-block-end-color-001.html
and similar tests.

The example of the problem, test above would compare two renderings:

Reference (kind-of-widget-fallback-input-button-border-block-end-color-001-expected.html)
    <input id="button-input" type="button" value="input-button">

vs:

Test (kind-of-widget-fallback-input-button-border-block-end-color-001.html)
    <input id="button-input" type="button" value="input-button">

    const prop = "border-block-end-color";
    for (const el of elements) {
        el.style.setProperty(prop, getComputedStyle(el).getPropertyValue(prop));
    }

The BorderPainter would compare the colors of each border, and run the
fast path if all the colors match.

For the reference, the fast path was taken.

For the test, because of the BorderPainter bug, the colors would not match. The
BorderPainter would take slow path that would set up complex clip path
with antialiasing.

Because the Recorder::clipToPath would incorrectly skip setting the antialiasing
state, two corner pixels of the rounded rect would be included in the rendering.
This would match the unclipped fast path rendering of the reference.

Once the clipToPath() bug was fixed, it would show two-pixel difference due
to the clipToPath being correctly cliping antialiased path vs the reference
using the fast path, unclipped.

Fix by fixing BorderPainter to compare the border colors always without
the "semantic" bit.

This makes the test take the same fast path as the reference, and thus the
rendering is identical.

Also in a best-effort manner, fix one other color compare case to ignore
the semantic bit. It is not consistent that one code-path would ignore
the semantic bit while others would be continue using it.

* Source/WebCore/platform/graphics/displaylists/DisplayListRecorder.cpp:
(WebCore::DisplayList::Recorder::clip):
(WebCore::DisplayList::Recorder::clipOut):
(WebCore::DisplayList::Recorder::clipPath):
(WebCore::DisplayList::Recorder::clipToImageBuffer):
(WebCore::DisplayList::Recorder::paintFrameForMedia):
(WebCore::DisplayList::Recorder::paintVideoFrame):
* Tools/TestWebKitAPI/Tests/WebCore/DisplayListRecorderTests.cpp:
(TestWebKitAPI::createTestPath):
(TestWebKitAPI::createTestImageBuffer):
(TestWebKitAPI::checkEqualState):
* Source/WebCore/rendering/BorderPainter.cpp:
 (WebCore::BorderPainter::paintSides):
 (WebCore::BorderPainter::paintTranslucentBorderSides):

Canonical link: https://commits.webkit.org/261618@main
